# Notebook 1: Chapter 2 - Part 1

We will not cover Chapter 2 in quite the same order as the text. This notebook is meant to guide you through part of Chapter reviewing or learning some Python. We will also explore many examples of the visualization capabilities. These will serve as examples and starting points for some of your work. 

I advise using this in parallel with the text. Work through the numbered exercises. Take enough time to think about the code, note the format, appreciate the possibilities. In places throughout the notebook, I will draw your attention to certain observations that I want you to make.

Each of you will work through the material at a somewhat different pace. My hope is that, in this format, I can point out certain information to each of you as you arrive at that point in your learning.

**Goals**
This exercise will serve several purposes including
1. Review or familiarize yourself with Python
2. Explore an array of tools we will be using or that are available such as
   a. visualization capabilities of *vpython* and *matplotlib*.
   b. SymPy module for symbolic mathematical calculations
   c. animation capabilities
3. Familiarize yourself with the text format and location of resources (online code, etc.)
4. Familiarize yourself with Jupyter Notebooks in case those are new (practice shortcut keys)
5. Do some first exercises both coding to solve a problem and interpreting code and pseudocode.

As a group you undoubtedly have a wide range of prior experience. I hope that everyone benefit from the list above. Also, I hope at the end of the exercise we have some similar, common understanding of the tools needed for the course.

This notebook will have you work through the text. That is really how it is structured. As you do there will be exercise embeded. I may also supplement at times with activities from other texts. 

This notebook will have you
1. Work through section 2.3 to review python basics.
2. Work through section 2.4.1. (We will come back to the other parts of 2.4.)
3. Work through Section 2.5 to learn some of the visualization and animation capabilities
4. Complete some exercises along the way to reinforce your understanding.

## 2.3 Python Mini Tutorial

We begin Section 2.3 by learning how to define (def) a function in Python. The first line 
<p style="text-align:center;">def <function name>(arguments):"</p> always has this format. Everything after that is indented four spaces (or more if it is inside the while loop structure). The code performs the calculation or action that you specify. At the end this function returns something, a numerical value in this case. Once the function is defined we can call it anytime later in the program and do the same calculation without rewriting all the code. 
    
**1. Enter the code for the Defunct function. Observe the error in the code the textbook gave you. This is a common mistake and easy to fix. Fix it by adding the line x=2 just before calling Defunct(x, 3).**

In [2]:
def Defunct(x, j): # Defines the function
    i = 1
    max = 10
    while (i < max):
        print(i)
        i = i + 1
    return i*x**j

x = 2
Defunct(x,3) # Calls the function

1
2
3
4
5
6
7
8
9


80

It's useful to avoid lines that are too long so we can have screens open side by side. A \ at the end of a line tells the computer to continue on the next line. 

**2. Edit the following line using a \ in appropriate places to display the following long line on three separate lines.**

In [4]:
x = 10
print ( 1 + x - x**2 + x**3 - x**4 + x**5 - x**6 \
       + x**7 - x**8 + \
       x**9 - x**10 + x**11 - x**12 \
       + x**13 - x**14 + x**15 )

909090909090911


Python has built-in math functions you can access by importing the math package as follows. 

**3. Note the functions listed in the text (or look them up online), and constants math.e, math.pi, math.inf, and math.nan. In the following cell, use pring statements and determine if the math.log() function is a base e or base 10 logrithm. Demonstrate how to obtain both base e and base 10 logrithms.**

In the following cell notice the generic log() function is natural log, ln(). Add a couple lines where you print a couple of these. Remember to put "math.". What's log(math.e)?**

In [12]:
import math
math.log(math.e)

1.0

In Python, we can assign a value to almost any variable name, EXCEPT a few keywords listed in Section 2.3.2 that are reserved. 

**4. Does False = True, import = 1, or continue = 'Please stop already!'?  In the next cell, I dare you to try...**

In [15]:
False = True

SyntaxError: cannot assign to False (2778346704.py, line 1)

In [16]:
import = 1

SyntaxError: invalid syntax (3407073972.py, line 1)

In [17]:
continue = 'Please stop already!

SyntaxError: unterminated string literal (detected at line 1) (1157435483.py, line 1)

Python variables can be integers (int), floating point numbers (float) complex numbers (complex), booleans (bool) or strings (str). Most languages have floats (single precision numbers) and doubles (double precision numbers). In python, floats already have double precision. Next week we'll learn more about the finite precision of numbers and problems this can cause.

In some languages if you divide two integers, you get an integer output (it drops anything after the decimal point. Others convert to a float (decimal) number. 

**5. What does python do? The author said his computer printed 3/6 as 0 rather than 0.5. Mine printed 0.5. Try printing the division case in the text (6/3, 3/6 and 3./6).**


In [22]:
6/3

2.0

In [39]:
x = 3.
y = 6

x*y

18.0

In [23]:
3./6

0.5

In most programming languages (C, Fortran, Java, etc) you must specify the type of variable when you create it. In C 

int x;
double y;
x = 10.0;

creates a variable x that is an integer and a variable y that is a double precision number. This tells the computer how much memory to store for that type of number. If you assigned a decimal value to an integer (x = 1.23) it will cause an error. There's not enough memory reserved in x. 

Python can tell the type and will switch types if needed. 

**6. Type in the code from the textbook to see what happens when you multiply or divide numbers that are integers and floats. The decimal point tells it the number is a float.**

In [27]:
4.000*6.000

24.0

In [28]:
4*6

24

In [31]:
16.000/3.000

5.333333333333333

In [32]:
16/3

5.333333333333333

Importing packages gives us access to enormous amounts of pre-programmed code from basic math functions in *math* or *numpi*, to graphics capabilities in *matplotlib*, to machine learning and AI tools in *scikitlearn*.

**7. In the next few cells, import the math package and see how complex numbers work. You can enter use the code in the text that describes complex numbers but feel free to experiment too. You will likely have to add some print statements to see the values.**

In [41]:
2j + 3

(3+2j)

**Continue exploring complex numbers in polar coordinates. Do you get an error? You can import the *cmath* package. Do you also get an error if you use the phase() function? It might not know where to look for that function. In that case, cmath.phase() tells it the function phase() belongs to the *cmath* package.**

In [42]:
import cmath

In [47]:
cmath.phase(2j+3)

0.5880026035475675

**Let's try playing with the polar representations...**

In [48]:
#Create a complex number in a cartesian representation, z = x + iy
import cmath

z = complex(1., 1.)
print(z)
zpolar = cmath.polar(z)
print(zpolar)
zrect = cmath.rect(zpolar[0],zpolar[1])
print(zrect)

(1+1j)
(1.4142135623730951, 0.7853981633974483)
(1.0000000000000002+1j)


On my computer some rounding or machine precision error showed up. I started with 1+1j and, after converting to polar representation and back to a Cartesian representation, ended with 1.0000000000000002+1.0000000000000002j. At least on my machine. Your computer or Python version may possibly be different.

**Strings**

Let's explore string data types and techniques to deal with these. Strings are just lists of characters. In other languages manipulating strings can be a laborious process. Python makes dealing with strings much simpler, in particular due to a technique called **slicing**, which can extract parts of a string very simply.

**8.(a) Type in the code from the text and try the different cases of single, double and triple quotes.**

**(b) Then print different array elements (0, 1, -1, -2). Play around until you understand what it's doing.**
    
**(c) Then enter the code to slice the array. Try changing the limits from [0:3] to something else and see how it works. What if you don't start at 0?**

In [53]:
S = "A string using double quotes"
S = "A string using single quotes"
S = 'It\'s possible to escape a quote'
""" From "Computational Physics"
Problem Solving with Python """ # Double within triple quotes

' From "Computational Physics"\nProblem Solving with Python '

In [50]:
S[0]

'I'

In [57]:
S[1:5]

"t's "

In arrays (and similar structures) the position of each element is assigned a number for its position, beginning with 0 in Python and C. In some programming languages the numbering starts with 1. Notice when we refer to a "slice" of the array S above [1:4] it prints the letters in positions 1, 2, and 3, but not 4 so print([1:N]) will print the values in positions from 1 to N-1. (In otherwords, the last element in the array is in position N-1 because we started counting from 0.)

**Math Operators**

If you are unfamiliar with the % and ** operators, % is the modulus operator. It returns the remainder a fraction. And ** is quick way to do an exponential. 5**2 = 25. Many languages don't have this option for exponentials.

**Conditional _if_ statements**

Most of the time our code will have to make decisions depending on certain conditions being met. If the conditions are not met, the program might do something else. We can us **_if_** or **_if_**-**_else_** or **_if_**-**_elif_**-**_elif_**-...-**_else_** statements to accomplish this.

Conditional statements are used to determine if a block of code is executed or not. For example

```
if condition:
   <if-block>
elif condition 1:
    <elif-block1>
elif condition 2:
    <elif-block2>
...
else:
    <else-block>
```

where the code that your program runs depends on which condition is **_true_**. You can have however many *elif* statements as you want or need. Note the : and indentation of the blocks of code by 4 spaces. These are important!

**9. The following are just a few examples of conditional statements, which return a boolean value *true* or *false*. Add one or two of your own.**

In [4]:
print(1 < 2)
print(2 < 1)
print(2 != 1)
print(2 == 1)

True
False
True
False


**Loops**

Most of the time we also need to perform a calculation or call a function many, many times in our programs. This is one great advantage of computers of course. We can use loops to do these repeated actions. The two main loops are *for* loops and *while* loops.

**10. Run the following two examples of loops. Change the parameters in the range function *for* example or the maximum range.**

In [5]:
for index in range(1, 3):
    print(index)

1
2


In [6]:
counter = 1
max = 10
while counter < max:
    print(counter)
    counter += 1

1
2
3
4
5
6
7
8
9


Note how the counter increments each time through the while loop until the condition is met. You don't have to increment by 1.

**11. Try incrementing by 2 or 3 each time. Copy and paste the *while* loop code into the next cell and try adding 2 each time.** 

**= does not mean equal**

Note that counter += 1 also is shorthand for counter = counter + 1. This second statement is a subtle but very important point. The = sign in a program is an *assignment* operator NOT an *equal* in the mathemetical sense. The code takes whatever is on the right side of the equal sign and replaces the variable on the left with that value. So counter = counter + 1, takes the old value of counter, adds one, and only then updates the variable counter with the new value.

### 2.3.4 Python Lists as Arrays
Let's explore Python *lists* aand *tuples* a bit. First, lists. Lists do not have to contain only numbers but we will often use them this way. Lists are similar to what other languages call arrays. We will explore "array* data types from the *NumPy* package later in the course.

**12. Enter all the code in section 2.3.4 regarding lists in the next block.**

Notice that you can modify individual elements of the list. In this case, the 0th element is changed from 1 to 5. 

A similar structure in Python is a *tuple*. They are created using parentheses rather than square breckets. They are slightly different.

**13. Enter the code for a *tuple* from the text example. Print the third element and the entire _tuple_.**

**Now try changing a single element, the first one, just like we did for the list and see what happens.**

Lists are *mutable* objects (they can be changed). Tuples are *immutable* objects (they cannot be changed.) 

The difference between creating these two objects is the *lists* are created with () and *tuples* with [].

A super convenient feature of Python is that you do not have to specify the size or type of an array or list in Python. Multi-dimensional arrays can be created by making a list of lists. For example,


In [7]:
L = [[1, 2], [3, 4], [5, 6]]      # A list of lists
print(L)                         # Print the whole list
print(L[0])                      # Print the first element

[[1, 2], [3, 4], [5, 6]]
[1, 2]


**14. In the following cell, create a list or two and play with some of the operations in the tables. Especially try the *append()* opperation. This adds new elements to the end of the list. An extraordinaraly convenient task that most other languages don't allow! Also try printing the index (position in list) of one of the specific elements of the list. Determine the length of a list.**

### 2.3.5 Python I/O

Just in case you are familiar with Python 2, but not Python 3, an important difference is the print function. The Python 2 method below throws an error.

```
print 'Hello, World!'     #Python 2
print('Hello, World!')    # Python 3
```

**14. Enter the simple code to input and print your name and age. Is astr() correct, or should it say str()?**

Interactive input is useful sometimes, though often data is read from a file. We will see how to do this in the exercises in the next section. For now let's explore formatting. When we print values it is often useful to limit the output to a specific number of decimal places. Let's see how to format output. The following cell has two print statements from the text, including apparent typos. 

**15. Edit these both so that they print the correct output.** Debugging can be frustrating at first but it is an effective way to understand the subtle points of correct formatting.

In [8]:
import math
x = 12.345132432
age = 39

print("x=%6.3f,  Pi=%9.6f,  Age=%d \n") % (x, math.pi, age)
print("x=%6.3f, %(x), "Pi=%9.6f," %(math.pi), "Age=%d "%(age)," \n)

SyntaxError: invalid decimal literal (776832746.py, line 6)

Notice for the floats the format %6.3f specifies the number of characters, 6, and the number of places after the decimal point, 3. This kind of formatting is very common in other languages too. Pleaase note that Python has other ways to format output which you may have learned. But the text uses this we will encounter this throughout the text. Best to be familiar with it.

One feature of the textbook is the "Code Listings" in the last section of each Chapter. Many, if not all of these are available on the textbook's website. I have a link on Blueline. It's preferable not to copy and paste because you want to practice entering code. It will allow you to do a bit of debugging of simple typos which is a noremal part of programming. But also we want to focus on the algorithms that are implemented so we don't necessarily want to start writing code from scratch everytime. 

Outside of class you are likely to look up code and documentation in standard or common packages and adapt these for your needs. 

**<span style="color:black;">16. Practice some of the escape and formatting characters listed in the textbook by entering the code from Directives.py in the next cell. Locate the code in Listing 2.12.</span>**

In [ ]:
# Directives.py illustrates escape and formatting characters
import sys
print("hello \n")
print("\t it's me")                           # tabulator
b = 73
print("decimal 73 as integer b = %d "%(b))    # for integer
print("as octal b = %o"%(b))                  # octal
print("as hexadecimal b = %x "%(b))           # works hexadecimal
print("learn \"Python\" ")                    # use of double quote symbol
print("shows a backslash \\")                 # use of \\
print('use of single \' quotes \' ')          # print single quotes

# Section 2.4 Programming Warmup

Read through this introduction to Section 2.4 and continue with Section 2.4.1 if you haven't read it already. You will enter Area.py below. 

We will skip parts 2.4.3 to 2.4.5 until next week.

### Section 2.4.1 Program Design

In this section we explore the general process of converting an idea for a computation into a piece of code the computer understands. In physics we don't often begin solving a problem by thinking in terms of equations. Through practice we think in terms of concepts, develop an idea about how a calculation should proceed, and finally beginning choosing the mathematical steps we need to solve the problem.

In computational physics we use a similar approach. We almost never begin thinking in terms of code. We start with a general idea of the main steps of the program, a so called pseudocode. Then we can work on the details. If it's a complex code it may take multiple iterations to develop and translate into actual code. We often have to test our code at different stages of development. Evaluating and revising the code is all part of the process.

The intro to Section 2.4 illustrates this with the Area.py program (also in Listing 2.10). Read through the intro and Section 2.4

**<span style="color:black;">17. Enter the Area.py program in the next cell and run it.</span>**

Writing a pseudocode and converting it to a program is useful for developing an idea. But what about reading other people's code and understanding it. In that case you very well may do the reverse process. Start with the code and write down what the code does in steps to understand how it works, how it is structured, or how to combine it with your code so they work together.

If your lucky, the code you want to use is well documented. There are plenty of comments explaining what it does. Unfortunately, this is not always the case. Let's use the AreaFormated.py program (Code Listing 2.11 or online) to explore this. 

**<span style="color:black;">18. Now enter AreaFormatted.py from Listing 2.11 in the next cell. You will also need to create the file *Name.dat* which the program will read. To creat the file, in the webpage with your directory/file list, click New at the top right and select text file. Name the file Name.dat. Enter your name and a number for the radius on one line. Save the file.</span>**

**Check if it created a file A.dat in the same directory as your program. What did it print? Create a Markdown cell below this and enter what is in A.dat or copy and paste.**

**NOTE: Anytime you turn in your work. Remember to upload any data files that the code needs to run.**

A lot of this code might be new depending on your experience with python. Review the code to see what it does. This code demonstrates how to open and close files and to read and write files. You might refer back to this as an example later. 

Notice this code has many comments. It explains what is happening at each step. You can quickly understand what it does without understanding every little detail. Imagine without comments. You would have to spend time looking through the details and working out what the author of the code is doing! This is why commenting code is so important and so useful. It saves a lot of time. And if someone else uses your code and you have commented it, you will get a lot fewer questions. More time saved!

**A main point here is comment your code a lot. And add Markdown cells explaining the code and output. This will be an important skill, developed as part of the computational essays you write throughout the semester.**


### 2.4.2 First Programming Steps
**19. Exercise: Complete section 2.4.2 using either the Area.py code (or AreaFormatted.py if you prefer) BUT rather than save different .py files, you can just adapt the code in the following cell.**

## 2.5 Python's Visualization Tools

While reading through Section 2.5, keep in mind the Code Listings in the textbook use an older module called *visual* that is obsolete. These codes have been translated to use *vpython* and are available online. The webpage was mentioned in the Preface and there is a link on Blueline. Select the option for VPython Codes.

*vpython* works in Jupiter notebooks and WebVpython.

Later, MatPlotLib is also explored to create different types of 2D and 3D visualizations as well.

**20. Read/work through each part of Section 2.5 in the text and enter your work/code in the corresponding sections below. Add a Markdown cell before code and explain what the code will do. Add a Markdown cell after and summarize something you have learned about the code or the output.**

**You should enter the code and familiarize yourself with how the code functions. You don't need to understand every detail but please take time to consider the format for creating different types of figures, how attributes of the figures are specified, how the axes are created and formatted, etc. These and many other attributes can be found online, on the *vpython* website for instance.** 

**_I mainly want you to see a range of tools available and have code you can refer back to for examples_.**

### 2.5.1 Visual (VPython)'s 2D Plots

You can find the *vpython* versions of the code on the webpage mentioned above.

**NOTE:** If you have issues making the figures appear, entering canvas() on a line after "from vpython import *" might work.



One useful feature of VPython figures is that they give you the coordinates if you hover the curser over it. That could be useful if you wanted to find the location of a maximum or minimum, or the location where two curves intersect, or the roots of an equation. 

VPython also has some nice 3D capabilities you will explore in the next section. This will also provide some more practice converting the code from the Visual to VPython. 


Next, I just want you to see an example of the 3D capabilities of *vpython*. I have adapted 3Dshapes.py in Code Listing 2.3 from the visual format to the vpython format. 

**21. Run the following code and see what you find.**

In [ ]:
# 3Dshapes.py: Some 3-D Shapes of VPython

from vpython import *
canvas()

graph1 = display(width=500, height=500, title='VPython 3-D Shapes', range=10)
sphere(pos=vector(0,0,0), radius=1, color=color.green)
sphere(pos=vector(0,1,-3), radius=1.5, color=color.red)
arrow(pos=vector(3,2,2), axis=vector(3,1,1), color=color.cyan)
cylinder(pos=vector(-3,-2,3), axis=vector(6,-1,5), color=color.yellow)
cone(pos=vector(-6,-6,0), axis=vector(-2,1,-0.5), radius=2, color=color.magenta)
helix(pos=vector(-5,5,-2), axis=vector(5,0,0), radius=2, thickness=0.3, color=color.orange)
ring(pos=vector(-6,1,0), axis=vector(1,1,1), radius=2, thickness=0.3,\
     color=vector(0.3,0.4,0.6))
box(pos=vector(5,-2,2), length=5, width=5, height=0.4, color=vector(0.4,0.8,0.2))
pyramid(pos=vector(2,5,2), size=vector(4,3,2), color=vector(0.7,0.7,0.2))
ellipsoid(pos=vector(-1,-7,1), axis=vector(2,1,3), length=4, height=2, witdth=5,\
          color=vector(0.1,0.9,0.8))

### (OPTIONAL) Converting visual to vpython code

In case you ever need to convert code that uses *visual* to the *vpython* format, here is a list of changes and tricks to make it work. There are a few other modifications that could be required depending on the particular code. These can be found on line, but this list of hints will get you started:

1. The code to import the package is different. Change "from visual.graph import *" to "from vpython import *"
3. Each time you want to draw graphs in a new figure, you must call the canvas() function. In EasyVisual.py, simply add the line "canvas()" after the first line importing the vpython package. If you don't do this, the graph will show up in the same cell where the last graph was drawn.
4. The gdisplay() function is now called graph(). Make sure to replace that.
5. If you try to run the code now, the first graph will not show up. That's because the gcurve color attribute is color.white. Also the default background is white. Simply change this to something else, like color.red. You can look up colors that are available by searching for VPython documentation online.
6. If you have encounter errors, you can often copy and paste into google to search for solutions. In later programs, there are likely other differences between the textbook's code that need adapted due to differences between visual and vpython.


### 2.5.2 Matplotlib's 2D Plots

Another common plotting package we will use is Matplotlib. Read through the text, if you haven't, and work through the examples below. The goal here is to be familiar with some of Matplotlib's capabilities and formatting as we will see this throughout the class as well.
When you enter the code, spend some time understanding what the different parts of the code do. How axes are created, labels, titles, the plot itself and different formatting features. If you want to find all the options possible, you can look these up inthe Matplotlib documentation online. You should practice looking up information!

**22. In the next cell, enter the code EasyMatPlot.py in the text or Code Listing 2.4 or on the website under Matplotlib Code. Spend a moment and play with some additional settings in Table 2.4.**

**23. Try the GradesMatplot.py code below. This example graphs data points asymmetric errorbars.**

Matplotlib allows us to construct more complex figures. The next example uses subplots to accomplish this.

**24. Use the code MatPlot2figs.py in Listing 2.6 or online to reproduce Figure 2.7.**

We often have to plot data points rather than continuous mathematical functions. Matplotlib can do this too.

**24. Run the following cell and explore how to do this using the code PondMatPlot.py from Listing 2.7.**

**NOTE:** The part of the code that generates the plot is what is relevant here. What is plotted will be explored later in Chapter 5. This code actually creates a graph you will see later in Chapter 5. **IMPORTANT:** If you get a syntax error related to the \pi or \sin escape sequences in the strings, you can add r in front of the string. For example, change \'2\$ \\ pi \$\' to 'r\'2\$ \\ pi\$\'.**

In [ ]:
# PondMatPlot.py:    Monte-Carlo integration via vonNeumann rejection

import numpy as np
import matplotlib.pyplot as plt

N = 100;    Npts = 3000;    analyt = np.pi**2
x1 = np.arange(0, 2*np.pi+2*np.pi/N, 2*np.pi/N)
xi = [];    yi = [];    xo = [];    yo = []

fig, ax = plt.subplots()
y1 = x1 * np.sin(x1)**2                              # Integrand
ax.plot(x1,y1, 'c', linewidth=4)
ax.set_xlim( (0, 2*np.pi) )
ax.set_ylim( (0, 5) )
ax.set_xticks( [0, np.pi, 2*np.pi] )
ax.set_xticklabels( ['0', r'$\pi$', r'2$\pi$'] )
ax.set_ylabel(r'$ f(x) = x\, \sin^2 x $', fontsize=20)
ax.set_xlabel('x', fontsize=20)
fig.patch.set_visible(False)

def fx(x):
    return x*np.sin(x)**2                 # Integrand

j = 0                                               # Inside curve counter
xx = 2.*np.pi * np.random.rand(Npts)                # 0 =< x <= 2pi
yy = 5*np.random.rand(Npts)                         # 0 =< y <= 5

for i in range(1, Npts):
    if (yy[i] <= fx(xx[i])):
        if (i <= 100):
            xi.append(xx[i])
        if (i <= 100):
            yi.append(yy[i])
            j += 1
    else:
        if (i <= 100):
            yo.append(yy[i])
        if (i <= 100):
            xo.append(xx[i])
    boxarea = 2. * np.pi * 5.                      # Box area
    area = boxarea*j/(Npts-1)                     # Area under curve
    ax.plot(xo,yo,'bo',markersize=3)
    ax.plot(xi,yi,'ro',markersize=3)
    ax.set_title('Answers: Analytic = %5.3f, MC = %5.3f'%(analyt,area))

plt.show()


### 2.5.3 Matplotlib's 3D Surface Plots

3D visualizations can be quite helpful to visualize multidimensional functions or data. Matplotlib can create 3D figures that you can rotate to get a better view of the figure. In this section we explore some examples that demonstrate these capabilities on Matplotlib using two pieces of example code. First look into a 

**25. In the next cell, use the code Simple3Dplot.py from Listing 2.8 (or online) to produce a 3D function. NOTE: _Replace the line ax = Axes3D(fig) with ax = fig.add_subplot(projection='3d')._ The default function should reproduce Figure 2.8. You can try playing with other functions if you'd like. "**

**Try changing "ax.plot_surface(X, Y, Z)", try adding the attributes "ax.plot_surface(X, Y, Z, cmap='viridis', alpha=0.5)".**


### AI as a learning tool

The previous example does not quite work as the authors have it. It needed the line "ax = fig.add_subplot(111, projection='3d')." The following cell was produced by Google's AI when I searched "How to use plot_surface and plot_wireframe in same figure matplotlib." That's how I discovered the "ax = fig.add_subplot(111, projection='3d') which made the plot show up. Using AI generated code might be helpful sometimes but it is important to understand what it does before you can use it. 

Searching for the line "ax = Axes3D(fig)" that we replaced, I discovered that it is deprecated, that is it is obsolete in the current version of Python, and should be replaced by the new line of code with the add_subplot() function instead. Just as we did. I went beyond just accepting the AI code and simply used it as a guide to point me to a deeper understanding of the code. I used it as a tool to enhance my understanding of Python.

We also learn from the AI code about the cmap and alpha attributes of plot_surface. You can look these up in the documentation for plot_surface online to understand what they do and how to use them. 

If you use AI code, it should be as a tool to understand more deeply. It can provide examples to investigate or to generate ideas. It might be helpful in conjunction with online documentation for various packages. Similarly, the webpages for documenting code often contain many, extremely useful examples to help get people started. If you use those *examples* as guides to adapt your code, you will soon be writing your own code when you need it. If you just copy and paste you will not learn, you will run into isues if you get a complex, specific task that AI can't handle it. And, a significant part of the grade will include understanding and explaining the code in your Computational Essays.

In [ ]:
import matplotlib.pyplot as plt
import numpy as np
from mpl_toolkits.mplot3d import Axes3D

# Create a figure and an axes object
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')

# Generate some data
X = np.arange(-5, 5, 0.25)
Y = np.arange(-5, 5, 0.25)
X, Y = np.meshgrid(X, Y)
R = np.sqrt(X**2 + Y**2)
Z = np.sin(R)

# Plot the surface
ax.plot_surface(X, Y, Z, cmap='viridis', alpha=0.5)

# Plot the wireframe
ax.plot_wireframe(X, Y, Z, color='black', linewidth=0.5)

# Set labels and title
ax.set_xlabel('X')
ax.set_ylabel('Y')
ax.set_zlabel('Z')
ax.set_title('Surface and Wireframe Plot')

# Show the plot
plt.show()

One more example before a final exercise. This example creates the 3D scatter plot.

**26. Enter the code from Scatter3dPlot.py in Listing 2.9 (or online), in the following cell.**

# 2.6 Plotting exercise

1. We encourage you to make your own plots and personalize them by trying out other commands and by including further options in the commands. The Matplotlib documentation is extensive and available on the Web. As an exercise, explore:
   
   a. how to zoom in and zoom out on sections of a plot, 

   b. how to save your plots to files in various formats, 

   c. ~~how to print up your graphs,~~ how to add text labels on your graphs (axis labels, title)

   d. the options available from the pull-down menus, 

   e. how to increase the space between subplots, 

   f. and how to rotate and scale the surfaces. 


a. Can change axis limits with plt.xlim(1,4) or plt.ylim(0, 0.5) for example. There are more complex ways, such as the inset axis, as demonstrated a couple cells down.
b. Files can be saved using plt.savefig("<Directory>/<filename>.png"), or .pdf, .gif, etc.
c. 
e. space between subplots can be changes using plt.subplots_adjust(wspace=0.4, hspace=0.4), where wspace controls the space horizontally (width-wise) and hspace controls the space vertically (height-wise).
f.



# Section 2.3.6 - Python's Algebraic Tools

I want you to be aware of these capabilities of Python for symbolic problem solving. This might be a useful tool in this class or other physics class too!

**Investigate SymPy capabilities. You can just work through the SymPy code in Section 2.3.6 of the text as an example.**

# Looking up resources and documentation

We will be using NumPy, Matplotlib, and VPython throughout the semester. It would be useful to know where to look for documentation and examples of use.

**Locate online documentation for NumPy, Matplotlib, and VPython. Enter the URLs here.**

# END OF NOTEBOOK